src/Features/Core/Portable/Completion/CompletionServiceWithProviders.ProviderSource.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    public abstract partial class CompletionServiceWithProviders
+    {
+        private class ProviderSource : IEqualityComparer<ImmutableHashSet<string>>
+        {
+            private readonly CompletionServiceWithProviders _service;
+            private IEnumerable<Lazy<CompletionProvider, CompletionProviderMetadata>>? _importedProviders;
+
+            private readonly object _gate = new();
```

Comment:
can this become a SemaphoreSlim so we can asynchronously wait for this to eb initialized?